### ISSUE TITLE
[rfc] quantize_tensor op in pytorch/pytorch

### ISSUE THREAD

**Issue by vkuzo** [pytorch/pytorch#195481]
### 🚀 The feature, motivation and pitch

We propose a `torch.nn.functional.quantize_tensor` op in PyTorch for easy and fast single-kernel tensor quantization for popular recipes targeting `torch.nn.functional.scaled_mm` and `torch.nn.functional.scaled_grouped_mm`.  The frontend should cover popular quantization recipes on modern GPUs (mxfp8|6|4, nvfp4 with RHT and RS from https://arxiv.org/abs/2509.25149, fp8 blockwise from https://arxiv.org/abs/2412.19437) and be extensible enough to support future recipes.  The backend will be written in JIT DSLs, with a CPU fallback where possible.

Note that
* supporting every possible quantization recipe is a non-goal - we aim only to support a small set of recipes that are popular in the industry and efficient on modern hardware
* multi-level quantization recipes (such as nvfp4 with per-tensor scaling) assume that the outer scale has been calculated elsewhere. This API only calculates inner scales.
* Previous discussion is in https://github.com/pytorch/rfcs/issues/86 and https://github.com/pytorch/pytorch/issues/146414

# proposed frontend

These would be the user-friendly functions in the `torch.nn.functional` namespace.

```python
class InnerScaleCalc(StrEnum):
    # Defines how to get from (block of inputs, broadcasted chunks of outer scale)
    # to (scale, rcp_scale). Intentionally general to be able to cover
    # complicated use cases like 4over6
    RCEIL_E8M0 = "rceil_e8m0"
    NVFP4_E4M3 = "nvfp4_e4m3"

class RoundingMode(StrEnum):
    RTNE = "rtne"
    STOCHASTIC = "stochastic"

# dense
def quantize_tensor(
    # the input tensor
    input: Tensor,
    *,
    # the dtype of the calculated quantized data
    qdata_dtype: torch.dtype,
    # the inner scale calculation: how to get from a block of high precision inputs
    # and broadcasted outer_quant_scale to the inner scale and its reciprocal
    inner_scale_calc: InnerScaleCalc,
    # specifies the reduction granularity of inner scale, consistent with `F.scaled_mm`
    scaling_type: ScalingType | list[ScalingType],
    # specifies the optional scale swizzle, consistent with `F.scaled_mm`
    swizzle_type: SwizzleType = SwizzleType.NO_SWIZZLE,
    # specifies the rounding mode for quantized data, with RTNE and STOCHASTIC to be supported
    qdata_rounding_mode: RoundingMode = RoundingMode.RTNE,
    # specifies a pre-calculated outer quantization scale (such as for nvfp4 per-tensor or per-token)
    # note that this is the reciprocal of the outer scale that should be passed to scaled_mm|scaled_grouped_mm
    outer_quant_scale: Tensor | None = None,
    # specifies a source of randomness for stochastic rounding
    # Note: the final semantics of this argument are still TBD
    random_key_or_gen: Tensor | Generator | None = None,
    # specifies an RHT sign vector
    # Note: the final semantics of this argument are still TBD
    rht_sign_vector: Tensor | None = None,
    # returns (qdata, scale)
) -> tuple[Tensor, Tensor]: ...

# dense, dual
# same as quantize_tensor, but for quantizing along both k-dim vs m-dim in the same kernel
def quantize_tensor_dual(
    # most args same as quantize_tensor
    ..., 
    # some args modified for additional options only needed for dual quant
    # 1. outer_quant_scale can differ for k-dim vs m-dim
    outer_quant_scale: tuple[Tensor | None, Tensor | None] | None = None,
    # 2. rht can differ for k-dim vs m-dim
    rht_sign_vector: tuple[Tensor | None, Tensor | None] | None = None,
    # returns (qdata_k, scale_k, qdata_m, scale_m)
) -> tuple[Tensor, Tensor, Tensor, Tensor]

# grouped
def quantize_tensor_grouped(
    input: Tensor,
    offs: Tensor,
    # all other args same as quantize_tensor
    ...,
) -> tuple[Tensor, Tensor]

# grouped, dual
def quantize_tensor_grouped_dual(
    input: Tensor,
    offs: Tensor,
    # all other args same as quantize_tensor_dual
    ...,
) -> tuple[Tensor, Tensor, Tensor, Tensor]
```

## python vs c++ boundary

`F.quantize_tensor` will call into `torch._quantize_tensor` which has args normalized to be c++-boundary-friendly, same as is done with `F.scaled_mm` vs `torch._scaled_mm_v2`.

# order of operations

* v1: mxfp8
* v2: mxfp4, nvfp4, blockwise fp8.  The nvfp4 recipe specifically will require finalizing the design on randomness, RS, RHT.
* later: other formats, case by case

# examples

## mxfp8 dense quant

```python
# dim-k
qdata, scale = quantize_tensor(
    input,
    qdata_dtype=torch.float8_e4m3fn,
    inner_scale_calc=InnerScaleCalc.RCEIL_E8M0,
    scaling_type=ScalingType.BlockWise1x32,
    swizzle_type=SwizzleType.SWIZZLE_32_4_4,
)

# dim-m - pass input transposed and use dim-k
qdata_t, scale_t = quantize_tensor(
    input.t(),
    qdata_dtype=torch.float8_e4m3fn,
    inner_scale_calc=InnerScaleCalc.RCEIL_E8M0,
    scaling_type=ScalingType.BlockWise1x32,
    swizzle_type=SwizzleType.SWIZZLE_32_4_4,
)
```

## fp8 blockwise quant from https://arxiv.org/abs/2412.19437

This can be covered trivially by `ScalingType.BlockWise1x128|BlockWise128x128` and a new value of `InnerScaleCalc`.

## nvfp4 dual quant with RS and RHT

This is the quantization of grad_output according to the recipe in https://arxiv.org/abs/2509.25149

```python
go_sr_q_k, go_s_k, go_sr_rht_q_m, go_s_m = quantize_tensor_dual(
    grad_output,
    qdata_dtype=torch.float4_e2m1fn_x2,
    inner_scale_calc=InnerScaleCalc.NVFP4_E4M3,
    scaling_type=[ScalingType.BlockWise1x16, ScalingType.TensorWise],
    swizzle_type=SwizzleType.SWIZZLE_32_4_4,
    outer_quant_scale=(go_gs_k.reciprocal(), go_rht_g_s_m.reciprocal()),
    rht_sign_vector=(None, rht_sign_vector),
    qdata_rounding_mode=RoundingMode.STOCHASTIC,
    random_key_or_gen=key,
)
```

## 4over6

It's not yet clear if 4over6 (https://arxiv.org/abs/2512.02010) should be added to this API. If we decide to add it at a future time, it can be done as follows:

```python
class InnerScaleCalc(StrEnum):
    FOUR_OVER_SIX_NVFP4_E4M3 = "four_over_six_nvfp4_e4m3"
```

## square scaling factors for mxfp8|4 and nvfp4

We plan to eventually support square scaling factors for mxfp8|4 and nvfp4.  This requires additional design, as the `ScalingType` enum is written for `F.scaled_mm`, and a recipe such as "mxfp8 with 32x32 scaling" needs its scales expanded and swizzled to be usable by a gemm.  We plan add a flag such as `scaling_type_square_block_and_expand` to cover this, final design TBD.

## stochastic rounding for qdata

To cover the nvfp4 pretraining recipes we plan to support tile invariant stochastic rounding with both stateless and stateful RNG, and will defer finalizing the semantics until v2.  The arguments for user input for randomness (currently `random_key_or_gen`) may change.

## stochastic rounding for scale

If needed, this can go into the `InnerScaleCalc` enum

## random hadamard transform

To cover the nvfp4 pretraining recipes we plan to support random hadamard transforms. We need to finalize whether the API will take `rht_sign_vector` or `rht_tensor`.  This is punted until v2.

# alternatives considered

* 4 APIs for (dense|grouped, unidirectional|dual), vs a single API combining them all via arguments.  We are currently going with 4 APIs for clearer semantic meaning of each API.
* a unified API for all formats vs a separate API for each format.  We are currently going for a unified API to minimize framework maintenance cost, and covering all possible arg combinations is an explicit non-goal.  Unimplemented argument combinations will raise exceptions.
* dim-m quant specified by input shape (current) vs an enum (`QuantOrientation.TRANSPOSED`). We are currently going with an implicit "quantize along the last dim" to keep the # of arguments down.
* user-specified `f` lowered to a single kernel flex-attention style.  This is viable, but we think this could be a separate API, and we should still have a simple hardcoded `quantize_tensor` in core

**Comment by syed-ahmed** [comment]
CC: @rdspring1 @tbqh @wolfcomos @protonu
